### PR TITLE
Infer number of hyperoptimize workers

### LINKED ISSUE #1168: Hyperoptimize outputs train and test AUCs in two columns
## What and why
Our awesome new `hyperoptimize` recipe produces this helpful output:
```
2021-01-18 18:26:36 - INFO - hyperoptimizers:108 - Trial 7 completed with results:

recipe                                                                         train
output_tensors                     ['sts_death_weighted_loss_5_60_days_after_ecg'...
train_sts_death                                                             0.832156
train_sts_stroke                                                            0.668101
train_sts_renal_failure                                                     0.781069
train_sts_prolonged_ventilation                                             0.758749
train_sts_dsw_infection                                                     0.701376
train_sts_reoperation                                                       0.664980
train_sts_long_stay                                                         0.760649
test_sts_death                                                              0.771208
test_sts_stroke                                                             0.700575
test_sts_renal_failure                                                      0.761738
test_sts_prolonged_ventilation                                              0.718955
test_sts_dsw_infection                                                      0.324281
test_sts_reoperation                                                        0.606424
test_sts_long_stay                                                          0.793331
Name: 7, dtype: object
```

## Solution(s)
I think the output could be even more clear if the train and test tensor maps are in distinct columns.

### PR DESCRIPTION
**Pull request by StevenSong**
## Issues closed
closes #1168 

## Changelog
- [x] `--hyperoptimize_workers` is now optional and will use all the gpus available to the docker container set by `-d` flag to `run.sh`
- [x] results aren't logged, just refer to the final csv
- [x] added incremental update of result csv so unforseen errors dont result in heartbreak

### CODE REVIEW

File: ml4c3/hyperoptimizers.py [hunk 1]
```diff
@@ -59,6 +60,14 @@ def hyperoptimize(args: argparse.Namespace):
         f"Randomly selected {len(permutations)} hyperparameter combinations to try",
     )
 
+    # infer the number of hyperoptimize workers as the number of available gpus
+    if args.hyperoptimize_workers is None:
+        # cannot rely on built-in tensorflow methods because importing tensorflow
+        # makes all gpus visible and prevents setting visible devices within workers
+        n_gpus = str(subprocess.check_output(["nvidia-smi", "-L"])).count("UUID")
```
Comment (erikr) [inline ml4c3/hyperoptimizers.py:67]: tensorflow can be janky sometimes

**Review by erikr — COMMENTED** [review]
Can you clarify the scenario where a user would use `-d 0-3` but `--hyperoptimize_workers 2`? In other words, why do we still want `args.hyperoptimize_workers` to exist?

**Comment by StevenSong** [comment]
> Can you clarify the scenario where a user would use `-d 0-3` but `--hyperoptimize_workers 2`? In other words, why do we still want `args.hyperoptimize_workers` to exist?

for sklearn models, since you're not using gpus, you can set hyperoptimize_workers to any amount to just use cpu cores

so say for each model, you have 3 cpu cores backing the datasets and u want 1 core for the model itself, so each model takes 4 cores - then in hyperopt on mithril with 28 cores, you can set hyperoptimize_workers to 7 to use all 28 cores

**Comment by erikr** [comment]
So for CPU workflows, to parallelize 4 workers per "task" * 7 tasks = 28 cores, you would do:
```
--hyperoptimize_workers 7
--num_workers 4 
```

But for GPU workflows, `hyperoptimize_workers` is irrelevant, as we simply distribute tasks among GPUs available to the Docker container?

**Comment by StevenSong** [comment]
> So for CPU workflows, to parallelize 4 workers per "task" * 7 tasks = 28 cores, you would do:
> 
> ```
> --hyperoptimize_workers 7
> --num_workers 4 
> ```

I would amend to `--num_workers 3` as that saves 1 cpu for the model itself (not an issue for a gpu workflow as computation is done on gpu)

> But for GPU workflows, `hyperoptimize_workers` is irrelevant, as we simply distribute tasks among GPUs available to the Docker container?

yes, if using gpus, i see no reason for `-d 0,1,2` and `--hyperoptimize_workers 2`

**Comment by erikr** [comment]
> I would amend to `--num_workers` 3 as that saves 1 cpu for the model itself (not an issue for a gpu workflow as computation is done on gpu)
I'm confused what "1 cpu for the model itself" means. If each `sklearn` model is trained with `c` cores per `ml4c3`, that is the value it passes to the `n_jobs` parameter https://scikit-learn.org/stable/computing/parallelism.html.

Also, 3 workers * 7 tasks = 21 cores, yes?

**Comment by StevenSong** [comment]
> I'm confused what "1 cpu for the model itself" means. If each `sklearn` model is trained with `c` cores per `ml4c3`, that is the value it passes to the `n_jobs` parameter https://scikit-learn.org/stable/computing/parallelism.html.

where do we set `n_jobs` for sklearn models? I haven't been able to find it in the codebase yet - regardless though, `--num_workers` in our codebase in the context of model training defines the number of workers used to back each dataset - if there are 3 cores backing a dataset, then those 3 cores should be distinct from any cores also training the model and doing the computations

**Comment by StevenSong** [comment]
though perhaps it is safe to use all cores for data loading, I forgot that for sklearn models, the data loading and model fitting are called sequentially and not parallelized 

https://github.com/aguirre-lab/ml4c3/blob/b546ab6e6bcd08107ef6dd13a213613749a98ee9/ml4c3/models.py#L182-L202

where the entire dataset `X` is loaded into memory before `model.fit` - in that case, it might actually be desireable for `model.fit` to set `n_jobs` so that the actual modeling can utilize multiple cores as well

**Comment by erikr** [comment]
I think you're right, we don't use >1 core for training an individual sklearn model.

> I forgot that for sklearn models, the data loading and model fitting are called sequentially and not parallelized

This is why I was confused by the core count being further subdivided.

Out of scope for this PR, but we should update our sklearn pipeline to sequentially:

1. Use n workers to load data (dataset).
2. Use n workers to train the model.

**Comment by StevenSong** [comment]
> Out of scope for this PR, but we should update our sklearn pipeline to sequentially:
> 
> 1. Use n workers to load data (dataset).
> 2. Use n workers to train the model.
> 
> @raimonpv thoughts?

issue is up #1170

**Comment by StevenSong** [comment]
@erikr last update for this PR, added incremental update of results

**Comment by erikr** [comment]
> @erikr last update for this PR, added incremental update of results

Awesome! Can you please share some example output that demonstrates this change?

**Comment by StevenSong** [comment]
this is an intermediate `metrics-and-hyperparameters.csv` while hyeroptimization is still running but trial 0 and 1 have finished:

```
trial	recipe	conv_x	dense_blocks			spatial_dropout	dense_dropout	learning_rate	train_severe		test_severe
0	train	[32]	[128, 128, 64, 64, 32, 32]	0.4		0.3		0.0002		0.5948631295026124	0.5601968329140151
1	train	[35]	[64, 64, 64, 64, 64, 64]	0.5		0.4		5e-05		0.5355542600846982	0.5310941909539533
2	train	[48]	[128, 128, 64, 64, 32, 32]	0.4		0.2		5e-05		
3	train	[32]	[128, 128, 64, 64, 32, 32]	0.2		0.4		0.0001		
4	train	[40]	[112, 96, 80, 64, 48, 32]	0.5		0.3		0.001		
```

**Comment by erikr** [comment]
> in the event that user hits ctrl+c, machine restarted, or code errors out, this file still exists with some results

This functionality is helpful!

Minor requests:
1. Change column names `train_severe` to `auc_train_severe`, and `test_severe` to `auc_test_severe`.
2. Round down the decimals fo the reported values to 3 places; any further precision can be found in the log files of the specific trial.

**Review by erikr — APPROVED** [review]
Thank you for this update, I think `hyperoptimize` keeps getting better and better!

**Pull request merged**